File: python/stack/asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []
        for asteroid in asteroids:
            if asteroid > 0:
                stack.append(asteroid)
            else:
                broken = False
                while stack and (stack[-1] > 0 and asteroid < 0):
                    right = stack.pop()
                    if right >= abs(asteroid):
                        if right > abs(asteroid):
                            stack.append(right)
                        broken = True
                        break

                if not broken:
                    stack.append(asteroid)

        return stack

File: c++/stack/asteroid-collision.cc
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        std::vector<int> stack;
        
        for (int i = 0; i < asteroids.size(); ++i) {
            if (stack.empty() || stack.back() < 0 || asteroids[i] > 0) {
                stack.push_back(asteroids[i]);
                continue;
            }
            
            bool destroyed = false;
            while (!destroyed && !stack.empty() && stack.back() > 0) {
                if (stack.back() > - asteroids[i]) {
                    destroyed = true;
                    break;
                } else if (stack.back() == - asteroids[i]) {
                    destroyed = true;
                }
                stack.pop_back();
            }

            if (!destroyed) {
                stack.push_back(asteroids[i]);
            } 
        }

        return stack;
    }
};
